streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java

```diff
@@ -180,41 +217,225 @@ public void processInOrder(final K key, final V value, final long timestamp) {
+            final ValueAndTimestamp<Agg> valueAndTime = ValueAndTimestamp.make(rightWinAgg.value(), Math.max(rightWinAgg.timestamp(), timestamp));
+            windowStore.put(
+                key,
+                valueAndTime,
+                window.start());
+            tupleForwarder.maybeForward(
+                new Windowed<>(key, window),
+                rightWinAgg.value(),
+                null,
+                rightWinAgg.timestamp());        }
+
+        private void createPreviousRightWindow(final long windowStart,
+                                               final long currentRecordTimestamp,
+                                               final K key,
+                                               final V value,
+                                               final long closeTime) {
+            final TimeWindow window = new TimeWindow(windowStart, windowStart + windows.timeDifferenceMs());
+            final ValueAndTimestamp<Agg> valueAndTime = ValueAndTimestamp.make(initializer.apply(), currentRecordTimestamp);
+            putAndForward(window, valueAndTime, key, value, closeTime, currentRecordTimestamp);
+        }
+
+        private void createCurrentRecordLeftWindow(final Long previousRecordTimestamp,
+                                               final long timestamp,
+                                               final ValueAndTimestamp<Agg> leftWinAgg,
+                                               final K key,
+                                               final V value,
+                                               final long closeTime) {
+            final ValueAndTimestamp<Agg> valueAndTime;
+            // if there's a right window that the new record could create && previous record falls within left window -> new record's left window is not empty
+            if (previousRecordTimestamp != null && leftWindowNotEmpty(previousRecordTimestamp, timestamp)) {
+                valueAndTime = ValueAndTimestamp.make(leftWinAgg.value(), timestamp);
+            } else {
+                valueAndTime = ValueAndTimestamp.make(initializer.apply(), timestamp);
+            }
+            final TimeWindow window = new TimeWindow(timestamp - windows.timeDifferenceMs(), timestamp);
+            putAndForward(window, valueAndTime, key, value, closeTime, timestamp);
         }
 
-        private boolean isLeftWindow(final KeyValue<Windowed<K>, ValueAndTimestamp<Agg>> window) {
-            return window.key.window().end() == window.value.timestamp();
+        private boolean leftWindowNotEmpty(final long previousTimestamp, final long currentTimestamp) {
+            return currentTimestamp - windows.timeDifferenceMs() <= previousTimestamp;
+        }
+
+        // previous record's right window does not already exist and current record falls within previous record's right window
+        private boolean rightWindowNecessaryAndPossible(final Set<Long> windowStartTimes,
```

Comment:
Sorry but this method name continues to throw me off...the comment does a good job of reminding what the check actually does/means, but ideally the method name alone would do a reasonable job of that. What about `previousRecordRightWindowMustBeCreated` or `previousRecordRightWindowDoesNotExistAndIsNotEmpty` ? I know those are both super long, especially the 2nd option, but I personally think the 2nd option does the best job of providing the link between what the check actually does, and why we do it. It's better to be clear than concise (when you can't be both)